#32 Feature: create operator and set permissions

Niyonkuru1 opened the pull request:
### What does this PR do?
- CRUD operations for operator
- Setting them a role
### Description of Task to be completed?
As an admin, I want to create, update, and delete operators and set them a roles so that they perform the task they are assigned to only.

- The admin should be able to create, update, and delete a operator.
- The admin should be able to view a paginated list of operators in the system.

- Skeleton UIs/screen are/is shown when the page loads/is loading.
 - The page is responsive.
- Appropriate error messages are displayed on unsuccessful attempts
### How should this be manually tested?
`Clone the repo`

`checkout` to branch `ft-create-operator-and-set-permissions-32
`

pull the changes by `git pull`

install all packages by `running npm install`

start app by running `npm run start`

### Any background context you want to provide?
N/A

### What are the relevant pivotal tracker stories?
[Trello Board](https://trello.com/c/OlZ7LyxG/32-create-operator-set-permissions)

Screenshots (if appropriate)

Niyonkuru1 committed:
feat(landing-page): adding correct fonts to the page and shadow on the image.

- dashboard

[ Finishes #28 ]

Niyonkuru1 committed:
feat(landing-page): adding correct fonts to the page and shadow on the image.

- dashboard

[ Finishes #28 ]

Niyonkuru1 committed:
feat(landing-page): adding correct fonts to the page and shadow on the image.

- dashboard 2

[ Finishes #28 ]

Niyonkuru1 committed:
small typos callection

Niyonkuru1 committed:
feat(create-operator): creating operator and setting permission

-the adding edit functionality on the page by using rising state up function
-adding modals. including newFormComponent just to check and editFormComponent, adding the states to toggle them on and off, fixing the app to perform all crud operation on the operators
-connecting with the json web server and configuring the redux to use that servers by adding a thunk function which allow to make async call. AJAX Calls and request before dispatching an action
-Adding json server and connecting redux with the server
-Creating crud operations on the operator as an admin
-fetching datas from json servers using redux thunk functions
-Adding the edit and update form and implementing their functionality using react and redux

[ Work In Progress #32 ]

Niyonkuru1 committed:
adding operator page to the dashboard

Niyonkuru1 committed:
feat(create-operator): creating operator and setting permission

- adding nunito font to the page to fix the rendering bug
- Integrating and renaming different folders to reflect the general naming structure
- adding the operator component to the view and putting all of its stuff into admin-operator-stuff folders
- pulling from the branch of crud-bus to get the structures

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): creating operator responsive versions

- Targeting responsiveness of the four screen size 400pxpx, 768px, 1024px, 1280px
- Adding responsiveness to the edit form and create new operator form
- Using tailwind four media queries sm, md, lg, and xl.
- Modifying the grid properties for each screen sizes

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): almost finishing the responsiveness

- The responsiveness of the page and adding sidebar modal responsive with the mobile version

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): validation of the forms when creating operator by admin

-Using formik to validate the edit and create forms
-Removing the usage of redux from the project
-Removing json server functionality
-adding delete functionality through redux

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): creating operator and making testing with react library

- Mimicing the user experience during testing
- establishing two first test
- resolving the error resulted due to other already done test

[ Work In Progress #32 ]

Niyonkuru1 committed:
small typos

Niyonkuru1 committed:
feat(create-operator): creating operator and setting permission

Creating crud operations on the operator as an admin
Adding json server and connecting redux with the server
-fetching datas frmo json servers using redux thunk functions
-Adding the edit and update form and implementing their functionality using react and redux

[ Work In Progress #32 ]

Niyonkuru1 committed:
merging conflict resolving

Niyonkuru1 commented:
### What does this PR do?
- CRUD operations for operator
- Setting them a role
### Description of Task to be completed?
As an admin, I want to create, update, and delete operators and set them a roles so that they perform the task they are assigned to only.

- The admin should be able to create, update, and delete a operator.
- The admin should be able to view a paginated list of operators in the system.

- Skeleton UIs/screen are/is shown when the page loads/is loading.
 - The page is responsive.
- Appropriate error messages are displayed on unsuccessful attempts
### How should this be manually tested?
`Clone the repo`

`checkout` to branch `ft-create-operator-and-set-permissions-32
`

pull the changes by `git pull`

install all packages by `running npm install`

start app by running `npm run start`

### Any background context you want to provide?
N/A

### What are the relevant pivotal tracker stories?
[Trello Board](https://trello.com/c/OlZ7LyxG/32-create-operator-set-permissions)

Screenshots (if appropriate)

Niyonkuru1 committed:
feat(create-operator): Removin the conflicts from the merging and adding routes to the sidebar

- Adding the routing to the sidebar
- removing and correcting the problem with the styling
- merging the conflicts between the two merged branch

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): removing bugs which cause the

- adding the id to the newly created operator
- clean up the code
- lifting the state-up to allow the update and delete to be done on the parent-level --- at the *opator-admin* component

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): resolving bug on the modal on the mobile phone

- Resolving the responsiveness on the mobile version
- Fix the rendering of the correct icon on the phone modal

[ Finishes #32 ]

Niyonkuru1 committed:
Merge branch 'develop' of https://github.com/atlp-rwanda/phantom-fe-elite into merged-versions-demo

Niyonkuru1 committed:
Merge branch 'develop' of https://github.com/atlp-rwanda/phantom-fe-elite into merged-versions-demo

PrinceNiyonshuti commented on src/dashboard/Admin.js:
```diff
@@ -0,0 +1,55 @@
+// import TableRow from "./TableRow";
+// import { MdArrowDropDown } from "react-icons/md";
+// import { Link } from "react-router-dom";
+// import {
```
remember to remove the unused codes

PrinceNiyonshuti commented on src/dashboard/ErrorMessage.js:
```diff
@@ -0,0 +1,16 @@
+// import "../main.css"
+// import { ImSpinner9 } from "react-icons/im";
+
+// const ErrorMessage = () => {
```
the same here with unused codes

PrinceNiyonshuti commented on src/dashboard/FormModal.js:
```diff
@@ -0,0 +1,128 @@
+// import React from 'react'
+
+// const FormModal = () => {
+//   return (
```
same here and the rest of commented or unused codes

PrinceNiyonshuti reviewed (changes requested):
please remove the unused codes that are commented

Niyonkuru1 committed:
feat(create-operator): crud operations on the operator with json web server in action

- adding crud operations  functionality on the operator as an admin
- Adding json server and connecting db.json to the project
-fetching datas and updating them from json servers using axios requests

[ Work In Progress #32 ]

Niyonkuru1 committed:
Merge branch 'ft-create-operator-and-set-permissions-32' of https://github.com/atlp-rwanda/phantom-fe-elite into merged-versions-demo

Niyonkuru1 committed:
feat(create-operator): increasing test coverage

- increasing testing coverage
- testing the function of input,
- testing that header, forms, aside and main element are rendered in the page

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): Displaying the message for failing and successfull attempts while doing crud operations

- npm install react-hot-toast
- include the package in the view component --> Admin-operator.js
- writting a function responsible for displaying the async behaviours message
 when user attempting to perform any crud operations
 - raising the coverage
 - including the correct icon for attempting any actions

[ Work In Progress #32 ]

Niyonkuru1 committed:
feat(create-operator): Displaying the message for failing and successfull attempts while doing crud operations

- npm install react-hot-toast
- include the package in the view component --> Admin-operator.js
- writting a function responsible for displaying the async behaviours message
 when user attempting to perform any crud operations
 - raising the coverage
 - including the correct icon for attempting any actions
 - removing the select input and its functionality
 - refactoring the code

[ Work In Progress #32 ]

tysepa reviewed (approved):
working good but correct those conflict

Niyonkuru1 committed:
feat(create-operator): increasing test coverage with react testing library

- Increasing test coverage
- Adding mock function to fakeout their implementation

[ Finishes #32 ]

Niyonkuru1 committed:
Merge branch 'ft-create-operator-and-set-permissions-32' of https://github.com/atlp-rwanda/phantom-fe-elite into merged-versions-demo

Niyonkuru1 committed:
Merge branch 'develop' of https://github.com/atlp-rwanda/phantom-fe-elite into merged-versions-demo

PrinceNiyonshuti reviewed (approved):
looks good

thepatrickniyo commented on src/Routes/LandingPage.js:
```diff
@@ -0,0 +1,100 @@
+      <FooterBlock />
+    </>
+  );
+	// return (
+	// 	<>
+	// 		<NavBar />
+	// 		<main className="mb-auto flex-grow">
+	// 			<main className="py-8 md:pb-28 lg:pb-8 lg:mb-12 sm:pl-20 px-8 h-[70%]">
+	// 				<section
+	// 					className="flex
+  //       flex-col-reverse
+  //       sm:flex-row-reverse
+  //        items-center sm:justify-between
+  //        ">
+	// 					<img
+	// 						className="w-full sm:w-1/2"
+	// 						src={landing}
+	// 						alt=""
+	// 						role="image"
+	// 					/>
+	// 					<div>
+	// 						<h1 className="text-textBluePhant font-bold leading-9 text-3xl sm:text-4xl font-Montserrat">
```
Remember to remove comments for the best code quality

thepatrickniyo reviewed (changes requested):
Remove comments every where in your codebase

Niyonkuru1 committed:
feat(create-operator): Refactoring code to remove unused one and enhance form validation.

- Adding formik.onSubmit to the newForm component
- Adding Removing usused code and comments from components
- Adding correct color to the form buttons and hover effects on them
- Adding form validation using yup schema

[ Finishes #32 ]

Niyonkuru1 committed:
feat(create-operator): Refactoring code to remove unused one and enhance form validation.

- Adding formik.onSubmit to the newForm component
- Adding Removing usused code and comments from components
- Adding correct color to the form buttons and hover effects on them
- Adding form validation using yup schema

[ Finishes #32 ]

Niyonkuru1 committed:
feat(create-operator): Increasing test coverage

- Mocking axios get, post and delete request
- Adding axios.js in the __mock__ folder
- testing that edit form modal is popped up when edit button is clikced
- testing that when submit is clicked the put request is made
- changing the styles of the modal to better display error without flipping out

[ Finishes #32 ]
